docs/src/main/sphinx/sql/match-recognize.rst

```diff
@@ -0,0 +1,575 @@
+      DEFINE variable_definition [, ...]
+      )
+
+Description
+-----------
+
+The ``MATCH_RECOGNIZE`` clause is used to detect patterns across rows of a table. Patterns of interest are
+specified using row pattern syntax based on regular expressions. When the pattern is detected, requested
+information, specified as the row pattern measures, is retrieved from rows that match certain components
+of the pattern.
+
+Example
+-------
+
+In the following example, the pattern describes V-shape over the ``totalprice`` column. A match is found
+whenever orders made by a customer first decrease in price, and then increase past the starting point::
+
+    SELECT * FROM orders MATCH_RECOGNIZE(
+         PARTITION BY custkey
+         ORDER BY orderdate
+         MEASURES
+                  A.totalprice AS starting_price,
+                  LAST(B.totalprice) AS bottom_price,
+                  LAST(D.totalprice) AS top_price
+         ONE ROW PER MATCH
+         AFTER MATCH SKIP PAST LAST ROW
+         PATTERN (A B+ C+ D+)
+         SUBSET U = (B, C)
+         DEFINE
+                  B AS totalprice < PREV(totalprice),
+                  C AS totalprice > PREV(totalprice) AND totalprice <= A.totalprice,
+                  D AS totalprice > PREV(totalprice)
+         )
+
+Below, all subclauses of ``MATCH_RECOGNIZE`` clause are explained, using this example as illustration.
+
+Partitioning and ordering
+-------------------------
+
+The ``orders`` table, which is the input table of row pattern recognition, is partitioned by the column ``custkey``::
+
+    PARTITION BY custkey
+
+When partitioning is specified, pattern matching is performed independently on every partition. Otherwise,
+the whole input table constitutes a single partition. This is analogous to the semantics of ``PARTITION BY`` clause
+in :ref:`window specification<window_clause>`.
+
+The rows of the input table are ordered by the column ``orderdate``::
+
+    ORDER BY orderdate
+
+Although this clause is optional, it is generally useful to perform pattern matching on ordered data.
+Pattern matches a certain sequence of rows, so the order is meaningful part of the result.
+
+.. _row_pattern_measures:
+
+Row pattern measures
+--------------------
+
+The ``MEASURES`` clause allows to specify what information should be retrieved from a matched sequence of rows.
+
+.. code-block:: text
+
+    MEASURES measure_expression AS measure_name [, ...]
+
+A measure expression is a scalar expression whose value is computed based on a match. In the example, three
+row pattern measures are specified:
+
+``starting_price`` is the price in the first row of the match,
+
+``bottom_price`` is the lowest price (corresponding to the bottom of the "V" in the pattern),
+
+``top_price`` is the price in the last row of the match, which is also the highest price in the match, as required
+by the pattern.
+
+Measure expressions can refer to the columns of the input table. They also allow special syntax
```

Comment:
All of this should probably also go above the snippet

Reply:
I put it below the snippet purposefully. The snippet introduces the syntax of measure definition, and the following sentences refer to parts of the syntax (measure expression and measure name).